infra/base-images/base-builder/jcc/jcc2.go

```diff
@@ -0,0 +1,400 @@
+	contentsBytes, err := ioutil.ReadFile(srcFile)
+	if err != nil {
+		return err
+	}
+	contents := string(contentsBytes[:])
+	contents, err = CppifyHeaderIncludes(contents)
+	if err != nil {
+		return err
+	}
+	b := []byte(contents)
+	err = ioutil.WriteFile(srcFile, b, 0644)
+	return err
+}
+
+func CppifyHeaderIncludes(contents string) (string, error) {
+	shouldCppify, exists := os.LookupEnv("JCC_CPPIFY_PROJECT_HEADERS")
+	if !exists || strings.Compare(shouldCppify, "0") == 0 {
+		return contents, nil
+	}
+	if strings.Contains(contents, CppifyHeadersMagicString) {
+		return contents, nil
+	}
+	re := regexp.MustCompile(`\#include \"(?P<header>.+)\"\n`)
+	matches := re.FindAllStringSubmatch(contents, -1)
+	if len(matches) == 0 {
+		return "", nil // !!!
+	}
+	for i, match := range matches {
+		if i == 0 {
+			// So we don't cppify twice.
+			contents += CppifyHeadersMagicString
+		}
+		oldStr := match[0]
+		replacement := "extern \"C\" {\n#include \"" + match[1] + "\"\n}\n"
+		contents = strings.Replace(contents, oldStr, replacement, 1)
+		if strings.Compare(contents, "") == 0 {
+			panic("Failed to replace")
+		}
+	}
+	return contents, nil
+}
+
+func AppendStringToFile(filepath, new_content string) error {
+	// Appends |new_content| to the content of |filepath|.
+	file, err := os.OpenFile(filepath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	_, err = file.WriteString(new_content)
+	return err
+}
+
+func WriteStdErrOut(args []string, outstr string, errstr string) {
+	// Prints |outstr| to stdout, prints |errstr| to stderr, and saves |errstr| to err.log.
+	fmt.Print(outstr)
+	fmt.Fprint(os.Stderr, errstr)
+	// Record what compile args produced the error and the error itself in log file.
+	AppendStringToFile("/workspace/err.log", fmt.Sprintf("%s\n", args)+errstr)
+}
+
+func main() {
+	f, err := os.OpenFile("/tmp/jcc.log", os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		log.Println(err)
+	}
+	defer f.Close()
+	if _, err := f.WriteString(fmt.Sprintf("%s\n", os.Args)); err != nil {
+		log.Println(err)
+	}
+
+	args := os.Args[1:]
+	if args[0] == "unfreeze" {
+		fmt.Println("unfreeze")
+		unfreeze()
+	}
+	basename := filepath.Base(os.Args[0])
+	isCPP := basename == "clang++-jcc"
+	newArgs := append(args, "-w")
+
+	var bin string
+	if isCPP {
+		bin = "clang++"
+		newArgs = append(args, "-stdlib=libc++")
+	} else {
+		bin = "clang"
+	}
+	fullCmdArgs := append([]string{bin}, newArgs...)
+	if IsCompilingTarget(fullCmdArgs) {
+		WriteTargetArgsAndCommitImage(fullCmdArgs)
+		os.Exit(0)
+	}
+	retcode, out, errstr := Compile(bin, newArgs)
+	WriteStdErrOut(fullCmdArgs, out, errstr)
+	os.Exit(retcode)
+}
+
+type BuildCommand struct {
+	CWD string   `json:"CWD"`
+	CMD []string `json:"CMD"`
+}
+
+func WriteTargetArgsAndCommitImage(cmdline []string) {
+	fmt.Println("WRITE COMMAND")
+	f, _ := os.OpenFile("/out/statefile.json", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
```

Comment:
IIRC, `os.O_APPEND` caused an invalid JSON format during our hackathon, is it the same case here?